File: problem_solving/15_NumberLineJumps.rb
def kangaroo(x1, v1, x2, v2)
  if v1 <= v2
    return "NO"
  end 
  while x1 <= x2
    if x1 == x2
      return "YES"
    end
    x1 += v1
    x2 += v2
  end
  return "NO"
end

x1, v1, x2, v2 = 0, 2, 5, 3
p kangaroo(x1, v1, x2, v2)

# # ====================
# 問題:
#   二匹のカンガルーが一直線上に並び、それぞれのスタートポイントから同じ回数ジャンプし(カンガルー毎に飛距離が違う)、二匹のカンガルーが同じ地点に到達できる場合は "YES" を、追い越してしまった場合は "NO" を出力する。

# ポイント解説:
#   後ろから出発するカンガルーの飛距離が前にいるカンガルーの飛距離と同じか少ない場合は追いつくことができないので "NO"

#   ループ内でそれぞれの現在地に一回飛ぶ毎に飛距離をそれぞれ足していき、現在地が同じになった場合のみ "YES" を返す。

# # ==================== 
#   if jumping disctance of the first kangaroo is smaller than the second kangaroo's jumping distance, that means the first kangaroo will never reach to the other kangaroo. Return "NO" if this is the case.
  
#   In the while loop, add up jumping distance to the location until the first kangaroo's location is lower than the second kangaroo.
#   if the first kangaroo and the other kangaroo get to the same location, return "YES". Otherwise return "NO".

File: problem_solving/15_NumberLineJumps.js
function kangaroo(x1, v1, x2, v2) {
  let xd = x2 - x1;
  let vd = v2 - v1;
  if (xd === 0) {
    return "YES";
  } else if (vd === 0) {
    return "NO";
  } else if (vd > 0) {
    return "NO";
  } else if (xd % vd === 0) {
    return "YES";
  } else {
    return "NO";
  }
}

const x1 = 0;
const v1 = 2;
const x2 = 5;
const v2 = 3;

console.log(kangaroo(x1, v1, x2, v2));

// ====================
// Twitter @MUSACODE_JP さんより教えていただいた考え方で書きました。
// コメントありがとうございました！

// 問題:
//   ゲームのスコアを記録した配列を最初から順番に比較していく。
//   今までのスコアと比べて最高得点だった場合の最高得点記録更新回数をカウントする。
//   今までのスコアと比べて最低得点だった場合の最低得点記録更新回数をカウントする。
//   上記二つの数を返す問題。

// ポイント解説:
//   xd => 二匹のカンガルーの距離の差
//   vd => 二匹のカンガルーの飛距離の差

//   Line 4:
//     既に同じ位置にいる => "YES"

//   Line 6:
//     飛距離が同じ、つまり後のカンガルーが前のカンガルーに追いつくことは無い => "NO"

//   Line 8:
//     二匹のカンガルーの飛距離の差が0かマイナスの時、二匹の距離は開いていく => "NO"
  
//   Line 10:
//     二匹間の距離が二匹の飛距離の差で割り切れる時(ちょうど同じポイントに到達する) => "YES"

// ====================
//   xd => The distance between two kangaroos
//   vd => The difference of jump distance between two kangaroos

//   Line 4:
//     Already at the same point => "YES"

//   Line 6:
//     Both two kangaroos' jumping distances are the same, which means they will never get to the same point => "NO"

//   Line 8:
//     When the difference of both two kangaroos' jumping distances is zero or negative Number, distance of two kangaloos will be bigger => "NO"

//   Line 10:
//     When the distance of two kangaroos can be devided with the difference of two kangaroo's jumping distance, which means they will get to the same point => "YES"